src/puppetlabs/puppetdb/queue.clj

```diff
@@ -147,32 +160,62 @@
    (let [rx (metadata-rx (keys metadata-command->puppetdb-command))
          md-cmd->pdb-cmd metadata-command->puppetdb-command]
      (fn [s]
-       (when-let [[_ stamp md-command version certname] (re-matches rx s)]
-         (and certname
-              {:received (-> stamp
-                             Long/parseLong
-                             tcoerce/from-long
-                             kitchensink/timestamp)
-               :version (Long/parseLong version)
-               :command (get md-cmd->pdb-cmd md-command "unknown")
-               :certname certname}))))))
+       (when-let [[_ received-stamp producer-offset md-command version certname] (re-matches rx s)]
+         (let [received-time-long (Long/parseLong received-stamp)
+               producer-offset (and producer-offset (Long/parseLong producer-offset))]
+           (and certname
+                {:received (-> received-time-long
+                               tcoerce/from-long
+                               kitchensink/timestamp)
+                 :producer-ts (some-> producer-offset
+                                      (+ received-time-long)
+                                      tcoerce/from-long)
```

Comment:
Passing it through ks/timestmap means I lose the ability to compare it to other timestamps, as ks/timestamp returns a string. This was ok for received as I never needed to use it, just drag it around. I would like received to be an object as well. I wanted to make that switch, but figured it'd be best left to another ticket. I wrote up that refactor as PDB-3278